src/NzbDrone.Core/Messaging/Commands/CommandExecutor.cs

```diff
@@ -120,43 +68,74 @@ private dynamic GetCommand(string commandTypeName)
                 {
-                    MappedDiagnosticsContext.Set("CommandId", command.Id.ToString());
+                    MappedDiagnosticsContext.Set("CommandId", commandModel.Id.ToString());
                 }
 
-                handler.Execute((TCommand)command);
-                _trackCommands.Completed(command);
+                handler.Execute(command);
+                _commandService.Completed(commandModel);
             }
             catch (Exception e)
             {
-                _trackCommands.Failed(command, e);
+                _commandService.Failed(commandModel, e);
                 throw;
             }
             finally
             {
-                BroadcastCommandUpdate(command);
-                _eventAggregator.PublishEvent(new CommandExecutedEvent(command));
+                BroadcastCommandUpdate(commandModel);
+
+                _eventAggregator.PublishEvent(new CommandExecutedEvent(commandModel));
 
-                if (MappedDiagnosticsContext.Get("CommandId") == command.Id.ToString())
+                if (MappedDiagnosticsContext.Get("CommandId") == commandModel.Id.ToString())
                 {
                     MappedDiagnosticsContext.Remove("CommandId");
                 }
             }
 
-            _logger.Trace("{0} <- {1} [{2}]", command.GetType().Name, handler.GetType().Name, command.Runtime.ToString(""));
+            _logger.Trace("{0} <- {1} [{2}]", command.GetType().Name, handler.GetType().Name, commandModel.Duration.ToString());
         }
-
-
-        private void BroadcastCommandUpdate(Command command)
+        
+        private void BroadcastCommandUpdate(CommandModel command)
         {
-            if (command.SendUpdatesToClient)
+            if (command.Body.SendUpdatesToClient)
             {
                 _eventAggregator.PublishEvent(new CommandUpdatedEvent(command));
             }
         }
+
+        // TODO: We should use async await (once we get 4.5) or normal Task Continuations on Command processing to prevent blocking the TaskScheduler.
+        //       For now we use TaskCreationOptions 0x10, which is actually .net 4.5 HideScheduler.
+        //       This will detach the scheduler from the thread, causing new Task creating in the command to be executed on the ThreadPool, avoiding a deadlock.
+        //       Please note that the issue only shows itself on mono because since Microsoft .net implementation supports Task inlining on WaitAll.
+        public void Handle(ApplicationStartedEvent message)
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            if (Enum.IsDefined(typeof(TaskCreationOptions), (TaskCreationOptions)0x10))
+            {
+                for (int i = 0; i < THREAD_LIMIT; i++)
+                {
+                    _taskFactory.StartNew(ExecuteCommands, TaskCreationOptions.PreferFairness | (TaskCreationOptions)0x10 | TaskCreationOptions.LongRunning)
```

Comment:
Its not a grantee, its just a hint.
"It provides a hint to the TaskScheduler that oversubscription may be warranted."


https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskcreationoptions(v=vs.110).aspx

also, them dying is a big issue. if one of the threads gets aborted because of an exception, you are out that thread for the life of the app, you kill 3 threads and sonarr pretty much becomes a zombie and does nothing.